File: src/ticker.erl
-module(ticker).
-behavior(gen_server).

%% ticker API

-export([start_link/0, start_link/1, subscribe/1, unsubscribe/0]).

-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
         terminate/2, code_change/3]).

-define(SERVER, ?MODULE).
-define(DEFAULT_DELAY, 1000). % 1 second.

-record(state, {delay = ?DEFAULT_DELAY, pubsub = no_pid}).

%% API

%% Starts a ticker process with the default delay (1 second).
start_link() -> start_link(?DEFAULT_DELAY).

%% Starts a ticker process with a certain delay.
start_link(Delay) when Delay > 0 ->
    RegisteredName = {global, ?SERVER}, % 1 ticker process across entire cluster!
    Args = {delay, Delay},
    Options = [],
    gen_server:start_link(RegisteredName, ?MODULE, Args, Options). % TODO check what this gives when started on 2nd node => process result

%% Subscribes a process to the ticker process to start receiving ticks.
subscribe(SubFunction) when is_function(SubFunction, 1) ->
    gen_server:cast(?SERVER, {subscribe, SubFunction}).

%% Unsubscribes a process from the ticker process.
unsubscribe() -> gen_server:cast(?SERVER, {unsubscribe, self()}).

%% gen_server callback functions

init({delay, Delay}) ->
    process_flag(trap_exit, true),

    {ok, PubSub} = pubsub:start_link(),
    State = #state{delay = Delay, pubsub = PubSub},
    spawn_worker(Delay),
    {ok, State}.

handle_call(_Request, _From, State = #state{}) ->
    Reply = {error, not_supported},
    {reply, Reply, State}.

handle_cast({subscribe, SubFunction}, State = #state{pubsub = PubSub}) ->
    pubsub:add_sub(PubSub, SubFunction),
    {noreply, State};
handle_cast({unsubscribe, Pid}, State = #state{pubsub = PubSub}) ->
    pubsub:remove_sub(PubSub, Pid),
    {noreply, State};
handle_cast(time_expired, State = #state{delay = Delay, pubsub = PubSub}) ->
    pubsub:publish(PubSub, tick),
    spawn_worker(Delay),
    {noreply, State}.

handle_info(_Request, State = #state{}) ->
    {noreply, State}.

terminate(_Reason, _State) -> ok.

code_change(_OldVsn, State, _Extra) -> {ok, State}.

%% Helper functions

spawn_worker(Delay) ->
    spawn(fun() ->
        timer:sleep(Delay),
        gen_server:cast(?SERVER, time_expired)
    end).

File: lib/ticker.ex
defmodule Snake.Ticker do
  use GenServer
  alias Snake.PubSub

  @moduledoc """
  Process that emits a tick to a pubsub mechanism every X milliseconds.
  """

  @server {:global, __MODULE__}
  @default_delay 1000 # 1 second.

  defmodule State, do: defstruct delay: 0, pubsub: :no_pid

  # API
  
  @doc """
  Starts a ticker process with the default delay (1 second).
  """
  def start_link, do: start_link @default_delay

  @doc """
  Starts a ticker process with a certain delay in milliseconds.
  """
  def start_link(delay) when delay > 0 do
    args = {:delay, delay}
    GenServer.start_link(__MODULE__, args, [name: @server]) # TODO check what this gives when started on 2nd node => process result
  end

  @doc """
  Subscribes a process to the ticker process to start receiving ticks.
  """
  def subscribe(pid, sub_function) when is_function(sub_function, 1) do
    @server |> GenServer.cast {:subscribe, pid, sub_function}
  end

  @doc """
  Unsubscribes a process from the ticker process.
  """
  def unsubscribe(pid) when is_pid(pid) do
    @server |> GenServer.cast {:unsubscribe, pid}
  end

  # GenServer callbacks

  @doc false
  def init({:delay, delay}) do
    Process.flag(:trap_exit, true)

    {:ok, pubsub} = PubSub.start_link
    spawn_worker(delay)
    {:ok,  %State{delay: delay, pubsub: pubsub}}
  end

  @doc false
  def handle_cast({:subscribe, pid, sub_func}, 
                  state = %State{pubsub: pubsub}) do
    pubsub |> PubSub.add_sub(pid, sub_func)
    {:noreply, state}
  end
  def handle_cast({:unsubscribe, pid}, state = %State{pubsub: pubsub}) do
    pubsub |> PubSub.remove_sub(pid)
    {:noreply, state}
  end
  def handle_cast(:time_expired, 
                  state = %State{delay: delay, pubsub: pubsub}) do
    pubsub |> PubSub.publish :tick                
    spawn_worker(delay)
    {:noreply, state}
  end

  # Helper functions
  
  defp spawn_worker(delay) do
    spawn fn ->
      :timer.sleep(delay)
      @server |> GenServer.cast :time_expired
    end
  end
end
